src/compose/ComposeBox.js

```diff
@@ -217,6 +219,36 @@ class ComposeBox extends PureComponent<Props, State> {
     }
   };
 
+  replaceMessageTextAtRange = (text: string, beg: number, end: number) => {
+    const { message } = this.state;
+    this.setMessageInputValue(message.slice(0, beg) + text + message.slice(end, message.length));
+    return [beg, beg + text.length];
+  };
+
+  insertAttachment = async (attachments: [DocumentPickerResponse]) => {
+    this.setState(({ isUploading }) => ({
+      isUploading: !isUploading,
+    }));
+
+    const { _, auth } = this.props;
+    let [beg, end] = [undefined, this.state.selection.start];
+    for (let i = 0; i < attachments.length; i++) {
+      const fileName = attachments[i].name ?? `attachment-${i + 1}`;
+      [beg, end] = this.replaceMessageTextAtRange(`[Uploading ${fileName}...]()\n`, end, end);
+      try {
+        const response = await api.uploadFile(auth, attachments[i].uri, fileName);
+        [beg, end] = this.replaceMessageTextAtRange(`[${fileName}](${response.uri})\n\n`, beg, end);
```

Comment:
I think the overall sequence of what this code does, if there are several files to upload/attach, is going to be:
* We insert the first placeholder, and start uploading.
* Eventually the upload finishes. We edit that placeholder, insert the next one, and start uploading.
* Etc., until we edit the final placeholder.

And I guess that agrees with what I see in your helpful screencaps.

That seems like not quite the ideal sequence -- having the placeholders come in one by one may feel kind of glitchy.

I think a better sequence would be:
* We insert all the placeholders immediately. We start uploading the first file.
* Eventually the upload finishes. We edit that placeholder, and start uploading the next file.
* Etc., until we edit the final placeholder.

(I think this code already makes exactly the right choice in that it only attempts one upload at a time. Uploads are probably often big files, and we may often be on a slow network connection -- no sense hammering it with a lot of uploads at once.)

Reply:
So now I am doing it as you suggested (with a simple addition on failure):
- inserting all placeholders immediately.
- Search placeholders one by one and replace them with `linkText`.
- If any search fails, append the text at the very end of the message.
- If upload itself fails I am now replacing the placeholder with `[Uploading ${fileName} failed.]()`.

The last point is important to note since it differs from the web app and the previous implementation. I took this decision based on the fact that replacing it with nothing won't be an ideal choice, since it would be very easy to miss out that some uploads have failed (even with a toast message). a permanent replacement with `[Uploading ${fileName} failed.]()`. would be able to indicate a user (or the recipients) that there was more to send but it failed.

Reply:
That reasoning makes sense to me!

If the webapp's behavior is that when the upload fails, you just have the `[Uploading ${fileName}...]()` placeholder stick around indefinitely, then I think this `[Uploading ${fileName} failed.]()` behavior would be an improvement there, too. Perhaps mention it in #frontend on chat.zulip.org?